### ISSUE TITLE
[UX] Search settings: Add help text to explain Content Ranking

### ISSUE THREAD

**Issue by ghost** [backdrop/backdrop-issues#2743]
As per https://github.com/backdrop/backdrop-issues/issues/1852, I just noticed the same thing (no description of what ordering means/is for) on the Content Rankings on this page: /admin/config/search/settings - no description of what the numbers mean. I seem to recall only recently realising that I'd been doing them wrong on all my Drupal sites for years (putting them in the opposite order to what I thought).

Recommend adding help text to explain this.
---
PR by @klonos: https://github.com/backdrop/backdrop/pull/2385

**Comment by klonos** [comment]
...I have re-classified this as a bug, because a) users that are used to the Drupal/Backdrop ways of using "weights" are baffled when faced with the table, and b) the help text actually exists, but is not being rendered. See:

https://github.com/backdrop/backdrop/blob/1.x/core/modules/node/node.module#L1390

https://github.com/backdrop/backdrop/blob/1.x/core/modules/search/search.api.php#L117

**Comment by klonos** [comment]
PR up for review:

<img width="1187" alt="screen shot 2018-11-16 at 1 31 34 pm" src="https://user-images.githubusercontent.com/2423362/48594327-3fda9c80-e9a4-11e8-85d5-cd7d7ef6459a.png">

...while at it, I have:

- Made it so that the entire "Content ranking" fieldset is only rendered when there are actually influence factors available.
- Moved the "Logging" fieldset after "Content ranking". This way it is right before "Indexing", which makes more sense, and also "Content ranking" is right after the "Search items" fieldset, which also makes more sense 😄

**Comment by klonos** [comment]
... a little bit of an explanation on the approach of the PR:

So the help text already existed along with the entire "Content ranking" fieldset form element in `hook_search_admin()` in `search.api.php`. It was also duplicated by the `node_search_admin()` implementation in `node.module`. I have removed the entire fieldset from both these places, and I have moved in inside `search_admin_settings()` in `search.admin.inc`, where all the rest of the form elements that are rendered in that settings page reside. Both `hook_search_admin()` and `node_search_admin()` now return a `$factors` array instead of a `$form` array. There is no reason for this text (and the entire fieldset actually) to be repeated across every `*_search_admin` hook, and having it in a single place reduces the amount of strings we need to translate.

Also, the description itself was previously implemented as a separate `` element of type `#markup`. I have moved it to the `'#description'` property of the "Content ranking" fieldset.

> - Made it so that the entire "Content ranking" fieldset is only rendered when there are actually influence factors available.

You can get yourself in a situation where no factors are available, if on a vanilla installation you configure "Available search items" to only "User accounts" and also set that as the default (other wise you get a validation error):

<img width="1183" alt="screen shot 2018-11-16 at 1 59 56 pm" src="https://user-images.githubusercontent.com/2423362/48595300-2c313500-e9a8-11e8-9cbd-2f4ef96f52c9.png">

**Comment by opi** [comment]
Related https://github.com/backdrop/backdrop-issues/issues/3296

**Comment by herbdool** [comment]
Tested and it looks good.

**Comment by quicksketch** [comment]
Looks like it should work but there's an implementation detail that could be improved. Left a review in the PR: https://github.com/backdrop/backdrop/pull/2385#pullrequestreview-184459986

**Comment by BWPanda** [comment]
@klonos I tried applying the PR patch to my local, but it seems there have been significant-enough changes since to mean this needs a rebase.

**Comment by klonos** [comment]
Expected after almost 2 years 😅 ...let m revisit this then.

**Comment by klonos** [comment]
...the conflicts are created by #4197. I'll have to rethink/rework this.

**Comment by klonos** [comment]
>> - Made it so that the entire "Content ranking" fieldset is only rendered when there are actually influence factors available.
>
>You can get yourself in a situation where no factors are available, if on a vanilla installation you configure "Available search items" to only "User accounts" and also set that as the default (otherwise you get a validation error):

No need for that any more; things work as expected now.

**Comment by klonos** [comment]
I think that this entire issue may need to be closed, as there is now help text that explains how content ranking works, as well as what the influence numbers mean. Let me check if there's anything left from my PR that's worth implementing...

**Comment by klonos** [comment]
OK, here's what's left to be done:

- UX improvement: Move the "Logging" fieldset after "Content ranking". This way it is right before "Indexing", which makes more sense, and also "Content ranking" is right after the "Search items" fieldset, which also makes more sense.
- UX bug:
  - When the "Content" checkbox is enabled (default OOTB status), ticking/unticking it shows/hides the entire "Content settings" fieldset 👍 
  - Untick the "Content" checkbox and save -> ticking/unticking the checkbox now has no effect whatsoever 👎 (the "Content settings" fieldset is shown only after you save the form, which makes it a two-step process to add and configure content types/ranking)
- Although the "Content rankings" label and help text may seem as they are part of the rankings table below, they are actually a separate `'#type' => 'item'` thing, which has its own theme function (totally unnecessary IMO), with the help text being provided as a `'#help' => ...` property. This should either be a `'#description' => ...`, or the entire form item be of `'#type' => 'help'` w/o any special theme function.
- There's no reason for the help text to be duplicated in `hook_search_admin()` in `search.api.php`, as it adds translation overhead.

I think that I will close this issue and file either one or more separate issue(s), as the original scope and discussion of this issue is not relevant any more.

**Comment by klonos** [comment]
We don't have a "status - outdated" label, so closing this as a "works as designed" after #4197 got in.

I've filed #4760 as a follow-up.
